File: solutions/2026-07-18/python/2026-07-18-find-greatest-common-divisor-of-array.py
```python
# Problem: Find Greatest Common Divisor of Array
# LeetCode Link: https://leetcode.com/problems/find-greatest-common-divisor-of-array/
#
# Problem Summary:
# Find the greatest common divisor (GCD) of the smallest and largest elements in an integer array.
#
# Approach:
# 1. Find the minimum and maximum values in the input array `nums`.
# 2. Implement a GCD function (Euclidean algorithm is efficient).
# 3. Calculate and return the GCD of the minimum and maximum values found.
#
# Time Complexity:
# O(log(min(a, b))), where a is the minimum and b is the maximum element. This is due to the Euclidean algorithm for GCD.
# Finding min/max takes O(N) where N is the length of the array.
# Therefore, the overall time complexity is dominated by O(N).
#
# Space Complexity:
# O(1) - We only use a few variables to store min, max, and intermediate GCD calculations.

import math

class Solution:
    def findGCD(self, nums: list[int]) -> int:
        """
        Calculates the greatest common divisor of the smallest and largest numbers in an array.
        """

        # 1. Find the minimum and maximum values in the array.
        min_num = min(nums)
        max_num = max(nums)

        # 2. Implement a GCD function using math.gcd (built-in and efficient).
        # Alternatively, a custom Euclidean algorithm could be implemented:
        # def gcd(a, b):
        #     while b:
        #         a, b = b, a % b
        #     return a
        #
        # gcd_val = gcd(min_num, max_num)

        # 3. Calculate and return the GCD of the minimum and maximum values.
        gcd_val = math.gcd(min_num, max_num)

        return gcd_val

# Example Usage (for testing locally):
# sol = Solution()
# print(sol.findGCD([2, 5, 6, 9, 10]))  # Output: 2
# print(sol.findGCD([7, 5, 6, 8, 3]))  # Output: 1
# print(sol.findGCD([3, 3]))          # Output: 3
```

File: solutions/2026-07-18/java/2026-07-18-find-greatest-common-divisor-of-array.java
```java
// Problem Summary: Find the GCD of the smallest and largest numbers in an array.
// Link: https://leetcode.com/problems/find-greatest-common-divisor-of-array/
// Approach:
// 1. Find the minimum and maximum elements in the input array `nums`.
// 2. Implement a helper function `gcd` to calculate the greatest common divisor of two numbers using the Euclidean algorithm.
// 3. Return the GCD of the minimum and maximum elements found.
// Time Complexity: O(N + log(min(a, b))), where N is the length of the array.
//   - O(N) to find the minimum and maximum elements.
//   - O(log(min(a, b))) for the Euclidean algorithm to find GCD, where 'a' and 'b' are the min and max numbers.
// Space Complexity: O(1) as we only use a few variables to store min, max, and temporary values during GCD calculation.

class Solution {
    /**
     * Finds the greatest common divisor of the smallest and largest numbers in an array.
     * @param nums The input integer array.
     * @return The greatest common divisor of the smallest and largest numbers.
     */
    public int findGCD(int[] nums) {
        // Initialize min and max with the first element of the array.
        int minVal = nums[0];
        int maxVal = nums[0];

        // Iterate through the array to find the minimum and maximum values.
        for (int num : nums) {
            if (num < minVal) {
                minVal = num; // Update minimum value
            }
            if (num > maxVal) {
                maxVal = num; // Update maximum value
            }
        }

        // Calculate and return the GCD of the minimum and maximum values.
        return gcd(minVal, maxVal);
    }

    /**
     * Calculates the greatest common divisor (GCD) of two non-negative integers using the Euclidean algorithm.
     * The Euclidean algorithm is based on the principle that the greatest common divisor of two numbers
     * does not change if the larger number is replaced by its difference with the smaller number.
     * This process is repeated until one of the numbers becomes zero, at which point the other number is the GCD.
     *
     * @param a The first non-negative integer.
     * @param b The second non-negative integer.
     * @return The greatest common divisor of 'a' and 'b'.
     */
    private int gcd(int a, int b) {
        // Base case: if b is 0, then a is the GCD.
        if (b == 0) {
            return a;
        }
        // Recursive step: call gcd with b and the remainder of a divided by b.
        // This is equivalent to gcd(a, b) = gcd(b, a % b).
        return gcd(b, a % b);
    }
}
```